src/text_util.rs

```diff
@@ -12,10 +12,36 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::io;
+
+use crate::formatter::{FormatRecorder, Formatter};
+
 pub fn complete_newline(s: impl Into<String>) -> String {
     let mut s = s.into();
     if !s.is_empty() && !s.ends_with('\n') {
         s.push('\n');
     }
     s
 }
+
+/// Indents each line by the given prefix preserving labels.
+pub fn write_indented(
+    formatter: &mut dyn Formatter,
+    recorded_content: &FormatRecorder,
+    // prefix can be turned into FormatRecorder if templater gets runtime error type
```

Comment:
I didn't quite understand this comment. What's the connection between this and whether templater supports runtime errors? Could the caller evaluate the template into a FormatRecorder instead? Or is there another issue?

Reply:
I am still confused. Perhaps we are talking about different questions: I'm wondering "why not just use FormatRecorder everywhere" and your comment might address "why not use a function everywhere". See the diff below for what I meant.

Is that not flexible enough for some reason? If not, it would be nice if you explained that in the comment.

(By the way, all of these comments are optional; I don't mean to make explaining these things a blocker. I haven't yet approved this because I didn't yet go through the code as carefully as I'd like, but if you need it approved sooner, let me know. )


```diff
diff --git a/src/templater.rs b/src/templater.rs
index 3804a1a345...f1cc7acbd3 100644
--- a/src/templater.rs
+++ b/src/templater.rs
@@ -133,12 +133,9 @@
     fn format(&self, context: &C, formatter: &mut dyn Formatter) -> io::Result<()> {
         let mut recorder = FormatRecorder::new();
         self.content.format(context, &mut recorder)?;
-        text_util::write_indented(formatter, &recorder, |formatter| {
-            // If Template::format() returned a custom error type, this wouldn't compile.
-            // FormatRecorder will help there to discriminate template evaluation error from
-            // io::Error: call prefix.format(context, prefix_recorder)? out of this closure.
-            self.prefix.format(context, formatter)
-        })
+        let mut prefix_recorder = FormatRecorder::new();
+        self.prefix.format(context, &mut prefix_recorder)?;
+        text_util::write_indented(formatter, &recorder, &prefix_recorder)
     }
 }

diff --git a/src/text_util.rs b/src/text_util.rs
index 41ee89d6b3...cb78c1b2b9 100644
--- a/src/text_util.rs
+++ b/src/text_util.rs
@@ -28,7 +28,7 @@
 pub fn write_indented(
     formatter: &mut dyn Formatter,
     recorded_content: &FormatRecorder,
-    mut write_prefix: impl FnMut(&mut dyn Formatter) -> io::Result<()>,
+    recorded_prefix: &FormatRecorder,
 ) -> io::Result<()> {
     let mut new_line = true;
     recorded_content.replay_with(formatter, |formatter, data| {
@@ -36,7 +36,7 @@
             if new_line && line != b"\n" {
                 // Prefix inherits the current labels. This is implementation detail
                 // and may be fixed later.
-                write_prefix(formatter)?;
+                recorded_prefix.replay(formatter)?;
             }
             formatter.write_all(line)?;
             new_line = line.ends_with(b"\n");
```

Reply:
> I'm wondering "why not just use FormatRecorder everywhere"

Ah, okay, got it. The reason is that `FormatRecorder` is another indirection (and unfavorable abstraction) which isn't needed for the prefix. Suppose we do post-process template output in commands/mod.rs for example, it's better to not require the prefix recorded.

```rust
template.format(&commit, &mut recorder);
text_util::write_indented(formatter, &recorder, |formatter| formatter.write_str("    ")))
```

(In this example, `write_indented(formatter, &recorder, "    ")` is even better, though.)

Reply:
OK, i'd appreciate it if you reworded the comment along the lines of "We can't put recorder in a closure because..."

Reply:
Added code snippet as a comment. I think it should better describe the solution.